File: packages/agent-core/src/tools/builtin/planning/enter-plan-mode.ts
/**
 * EnterPlanModeTool — plan-mode entry tool.
 *
 * The LLM calls this tool to enter plan mode directly. Entering plan mode
 * does not require approval in any permission mode.
 */

import type { Agent } from '#agent';
import { z } from 'zod';

import { planModeEntryMessage } from '../../../agent/injection/plan-mode-contract';
import type { BuiltinTool } from '../../../agent/tool';
import type { ToolExecution } from '../../../loop/types';
import { toInputJsonSchema } from '../../support/input-schema';
import DESCRIPTION from './enter-plan-mode.md';

// ── Input schema ─────────────────────────────────────────────────────

export const EnterPlanModeInputSchema = z.object({}).strict();
export type EnterPlanModeInput = z.infer<typeof EnterPlanModeInputSchema>;

export class EnterPlanModeTool implements BuiltinTool<EnterPlanModeInput> {
  readonly name = 'EnterPlanMode' as const;
  readonly description: string = DESCRIPTION;
  readonly parameters: Record<string, unknown> = toInputJsonSchema(EnterPlanModeInputSchema);

  constructor(private readonly agent: Agent) {}

  resolveExecution(_args: EnterPlanModeInput): ToolExecution {
    return {
      description: 'Requesting to enter plan mode',
      approvalRule: this.name,
      execute: async () => {
        // Guard: already in a session mode. Name the ACTUAL active mode and its
        // exit tool so the model doesn't think it is in plan mode when it is
        // really in product/game-design/design.
        if (this.agent.sessionMode.isActive) {
          const kind = this.agent.sessionMode.kind;
          const active =
            kind === 'design' ? 'Design' :
            kind === 'product' ? 'Product' :
            kind === 'game-design' ? 'Game-design' :
            'Plan';
          const exitTool =
            kind === 'design' ? 'ExitDesignMode' :
            kind === 'product' ? 'ExitProductMode' :
            kind === 'game-design' ? 'ExitGameDesignMode' :
            'ExitPlanMode';
          return {
            isError: true,
            output: `${active} mode is already active. Use ${exitTool} when you are ready to exit ${active.toLowerCase()} mode; do not try to enter another mode on top of it.`,
          };
        }

        try {
          await this.agent.sessionMode.enter(undefined, undefined, undefined, 'plan');
        } catch (error) {
          const message = error instanceof Error ? error.message : 'Failed to enter plan mode.';
          return { isError: true, output: `Failed to enter plan mode: ${message}` };
        }

        this.agent.telemetry.track('plan_enter_resolved', { outcome: 'auto_approved' });
        return { output: planModeEntryMessage(this.agent.sessionMode.sessionModeFilePath) };
      },
    };
  }
}

File: rust-ody/crates/tools-rs/src/builtin/session_mode/enter_plan_mode.rs
use serde_json::Value;
use std::sync::Arc;

use crate::builtin::session_mode::{
    planning::plan_mode_entry_message, SessionModeKind, SessionModeProvider,
};
use crate::builtin::{
    BuiltinTool, ExecutableToolContext, ExecutableToolResult, ToolError, ToolExecution,
};

pub struct EnterPlanModeTool {
    provider: Arc<dyn SessionModeProvider>,
}

impl EnterPlanModeTool {
    pub fn new(provider: Arc<dyn SessionModeProvider>) -> Self {
        Self { provider }
    }
}

impl BuiltinTool for EnterPlanModeTool {
    fn name(&self) -> &str {
        "EnterPlanMode"
    }

    fn description(&self) -> &str {
        "Enter implementation-planning mode. Produces a step-by-step plan file."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {},
            "additionalProperties": false
        })
    }

    fn resolve_execution(&self, _args: Value) -> Result<ToolExecution, ToolError> {
        let provider = Arc::clone(&self.provider);
        Ok(ToolExecution {
            accesses: Default::default(),
            description: "Requesting to enter plan mode".into(),
            approval_rule: "EnterPlanMode".into(),
            matches_rule: None,
            display: None,
            execute: Box::new(move |_ctx: ExecutableToolContext| {
                let provider = Arc::clone(&provider);
                Box::pin(async move {
                    if provider.is_session_mode_active() {
                        let active = match provider.session_mode_kind() {
                            Some(SessionModeKind::Design) => "Design",
                            Some(SessionModeKind::OfficeHours) => "Office-hours",
                            Some(SessionModeKind::GameDesign) => "Game-design",
                            _ => "Plan",
                        };
                        let exit_tool = match provider.session_mode_kind() {
                            Some(SessionModeKind::Design) => "ExitDesignMode",
                            Some(SessionModeKind::OfficeHours) => "ExitOfficeHoursMode",
                            Some(SessionModeKind::GameDesign) => "ExitGameDesignMode",
                            _ => "ExitPlanMode",
                        };
                        return ExecutableToolResult::error_text(
                            format!(
                                "{} mode is already active. Use {} when you are ready to exit {} mode; do not try to enter another mode on top of it.",
                                active,
                                exit_tool,
                                active.to_lowercase()
                            ),
                            "session mode already active".into(),
                        );
                    }

                    if let Err(e) = provider.enter_session_mode(SessionModeKind::Plan).await {
                        return ExecutableToolResult::error_text(
                            format!("Failed to enter plan mode: {}", e),
                            "enter failed".into(),
                        );
                    }

                    provider.telemetry().track(
                        "plan_enter_resolved",
                        std::collections::HashMap::from([(
                            "outcome".into(),
                            "auto_approved".into(),
                        )]),
                    );

                    let msg = plan_mode_entry_message(provider.session_mode_file_path().as_deref());
                    ExecutableToolResult::ok_text(msg)
                })
            }),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::builtin::session_mode::tests::MockSessionModeProvider;

    #[tokio::test]
    async fn enter_plan_mode_succeeds_when_inactive() {
        let provider = Arc::new(MockSessionModeProvider::inactive());
        let tool = EnterPlanModeTool::new(provider.clone());
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;
        assert!(!result.is_error);
        assert!(result.to_text().contains("Plan mode is now active"));
        assert!(provider
            .entered
            .lock()
            .unwrap()
            .contains(&SessionModeKind::Plan));
    }

    #[tokio::test]
    async fn enter_plan_mode_fails_when_already_active() {
        let provider = Arc::new(MockSessionModeProvider::active(SessionModeKind::Design));
        let tool = EnterPlanModeTool::new(provider.clone());
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;
        assert!(result.is_error);
        assert!(result.to_text().contains("Design mode is already active"));
        assert!(result.to_text().contains("ExitDesignMode"));
    }
}
